### ISSUE TITLE
Fine-grained_NER 2018 model's memory conception is increase over time !

### ISSUE THREAD

**Issue by datix-vishu** [allenai/allennlp#4194]
Is there a way to solve this issue ?

**Comment by datix-vishu** [comment]
Is this happening because of vocab caching in ELMo ?

**Comment by datix-vishu** [comment]
I tried out a few ELMo based models and this is a common issue. If this is happening to everyone, then your demo will be also experiencing this memory increase for NER. So how are you handling this memory hike ?

**Comment by matt-gardner** [comment]
We'll need more detail than this in order to diagnose a problem - what does memory usage start at? How high does it go? Does it actually give you an OOM at some point?  When? Are you increasing the size of the instances you are passing through the model?

**Comment by datix-vishu** [comment]
When I load the fine-grained_NER model to the memory it takes around 800+MB. Then I pass one article at a time to the NER model. After around 8 or 10 articles the memory was at 1.7GB. This keeps increasing when a newer articles are fed to the model for prediction. Soon the VM (4GB ram ) crashes due to OOM . And yeah the length of the articles may differ from one another.

**Comment by matt-gardner** [comment]
We're not sure that this is actually a memory leak.  If you've only run 8 or 10 articles, the python interpreter may not have reached a steady state of memory allocation and freeing.  Can you run it longer in an environment with more memory to see if there actually is a leak here?

One potential issue that could be causing a leak is forgetting to put the model into eval mode, to turn off gradient computation (`model.eval()`).

**Comment by datix-vishu** [comment]
@matt-gardner Like this ?

from allennlp.predictors import Predictor
allennlp_model = Predictor.from_path('path to model')
allennlp_model.eval()
outputs = allennlp_model.predict(text)

**Comment by datix-vishu** [comment]
Is there any particular reason why Allennlp is slower than its competitors? I just used DeepPavlov NER model and I'm getting the results in 200ms.

**Comment by matt-gardner** [comment]
Oh, if you're loading the model from a predictor, we already call `.eval()` on the model, so that's not the issue.  Still, you might just have not gotten to a steady state of memory usage, and it looks like the problem is the VM getting killed, not python actually giving an OOM.  It looks like our demo runs in a VM that only requests 1GB of memory, which you can see [here](https://github.com/allenai/allennlp-demo/blob/f9dc7756c336a9495d26140f1d0797fe23933125/api/allennlp_demo/fine_grained_ner/.skiff/webapp.jsonnet#L16).  How long are the "articles" that you're passing to the model?  You should typically pass one sentence at a time to the model, not a full article.

Additionally, the goal of allennlp is not to provide an NER model.  It's to provide a library for helping researchers build NLP models.  We do not try to benchmark the performance of our demo models against anything else for speed, or even accuracy.  They are reference implementations of common research models, and that's it.
